src/native/http_request_response.c

```diff
@@ -311,6 +311,51 @@ void aws_java_http_stream_on_stream_destroy_fn(void *user_data) {
     /********** JNI ENV RELEASE **********/
 }
 
+void aws_java_http_stream_on_stream_metrics_fn(
+    struct aws_http_stream *stream,
+    const struct aws_http_stream_metrics *metrics,
+    void *user_data) {
+    struct http_stream_binding *binding = (struct http_stream_binding *)user_data;
+
+    /********** JNI ENV ACQUIRE **********/
+    JNIEnv *env = aws_jni_acquire_thread_env(binding->jvm);
+    if (env == NULL) {
+        /* If we can't get an environment, then the JVM is probably shutting down.  Don't crash. */
+        return;
+    }
+
+    /* Convert metrics to Java HttpStreamMetrics obj */
+    jobject jni_metrics = (*env)->NewObject(
+        env,
+        http_stream_metrics_properties.http_stream_metrics_class,
+        http_stream_metrics_properties.constructor_id,
+        (jlong)metrics->send_start_timestamp_ns,
+        (jlong)metrics->send_end_timestamp_ns,
+        (jlong)metrics->sending_duration_ns,
+        (jlong)metrics->receive_start_timestamp_ns,
+        (jlong)metrics->receive_end_timestamp_ns,
+        (jlong)metrics->receiving_duration_ns,
+
+        /* Stream IDs are 31-bit unsigned integers, which fits into Java's regular (signed) 32-bit int */
```

Comment:
is this true? Don't we have 32bit unsigned? Why not just use long?

Reply:
Well it is true that HTTP/2 stream IDs are 31-bit integers according to the spec ([RFC 7540 § 5.1.1](https://httpwg.org/specs/rfc7540.html#rfc.section.5.1.1)). It's also true that they're [treated as unsigned 32-bit by CRT](https://github.com/awslabs/aws-c-http/blob/6a1c157c20640a607102738909e89561a41e91e9/include/aws/http/request_response.h#L228).

I _could_ cast this to `long` but the value can never be outside the range of [0, 2³¹) so it fits inside a Java `int` without risk of loss/corruption. Given that it can fit into an `int` and `int`s are more idiomatic in Java than `long`s, that seems better.

Let me know if you feel strongly and I can switch this.